### PR TITLE
[react-redux] Update types for react-redux to support prop omission

### PR DESCRIPTION
**Pull request by clayne11**
The types were failing to remove props they injected from requirements
for the parents rendering them. By using Omit we can omit props that are
injected from the connect component so that we don't get errors during
parent rendering.

All of this is done through type inference so that end users don't have
to be aware that this is happening.

The `connect` component was also failing to pass through required props from the component it was wrapping. You can see in the test I added that `connect` will omit props that are passed in through the connect function but will still pass through the remaining props as required.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by thasner** [comment]
@clayne11 can you provide an example of the problem this change is intended to solve?

**Comment by clayne11** [comment]
I added one test in the PR but I can add more examples.

If you look at the test I added:

```javascript

namespace RemoveInjectedAndPassOnRest {
    interface TProps {
        showGlobalSpinner: boolean;
        foo: string;
    }
    class SpinnerClass extends React.Component<TProps & DispatchProp<any>, {}> {
        render() {
            return (<div />);
        }
    }

    export const Spinner = connect((state: any) => {
        return { showGlobalSpinner: true };
    })(SpinnerClass);

    <Spinner foo='bar' />
}
```

In the render test I added the prop `<Spinner foo='bar' />`. My test would fail without the `foo='bar'` since it passes through the `foo: string` prop requirement but filters out the `showGlobalSpinner` prop which is injected by the `connect` HOC. With the previous current implementation, the test would pass withough adding `foo='bar'` even though the prop requiremen ts aren't being met for the wrapped component.

If you need further clarification let me know.

File: types/react-redux/index.d.ts [hunk 1]
```diff
@@ -1,35 +1,50 @@
...
+// render.
+export type InferableComponentEnhancer<TInjectedProps> =
+    InferableComponentEnhancerWithProps<TInjectedProps, {}>
 
-interface ComponentMergeDecorator<TMergedProps, TOwnProps> {
-    (component: Component<TMergedProps>): ComponentClass<TOwnProps>;
-}
+// Injects default props and makes them optional. Will still pass through
+// the injected props if they are passed in during render.
+export type DefaultingInferableComponentEnhancer<TInjectedProps> =
```
Comment (thasner) [inline types/react-redux/index.d.ts]: is this used anywhere?

File: types/react-redux/index.d.ts [hunk 2]
```diff
@@ -50,73 +65,73 @@ interface ComponentMergeDecorator<TMergedProps, TOwnProps> {
  * @param mergeProps
  * @param options
  */
-export declare function connect(): ComponentDecoratorInfer<DispatchProp<any>>;
+export declare function connect(): InferableComponentEnhancer<DispatchProp<any>>;
 
 export declare function connect<TStateProps, no_dispatch, TOwnProps>(
     mapStateToProps: MapStateToPropsParam<TStateProps, TOwnProps>
-): ComponentDecorator<TStateProps, TOwnProps>;
+): InferableComponentEnhancerWithProps<TStateProps & DispatchProp<any>, TOwnProps>;
```
Comment (thasner) [inline types/react-redux/index.d.ts]: why are `DispatchProp<any>` added when `mapDispatchToProps` is undefined?

**Review by thasner — COMMENTED** [review]
thanks for the explanation

File: types/react-redux/index.d.ts (hunk 1 above)
Comment (clayne11) [inline types/react-redux/index.d.ts]: Nope good catch - I accidentally brought that over from other types I was working on. I'll remove it.

File: types/react-redux/index.d.ts (hunk 2 above)
Comment (clayne11) [inline types/react-redux/index.d.ts]: That's how `react-redux` works:

> If you do not supply your own mapDispatchToProps function or object full of action creators, the default mapDispatchToProps implementation just injects dispatch into your component’s props.

https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options

Comment (thasner) [inline types/react-redux/index.d.ts]: 👍

**Review by thasner — APPROVED** [review]
👍

**Review by thasner — CHANGES_REQUESTED** [review]
errr, waiting to remove the unused interface

**Review by thasner — APPROVED** [review]
👍

**Comment by amcasey** [comment]
@clayne11 Do you want to try to resolve the merge conflicts?

**Comment by clayne11** [comment]
That merge is a bit broken - I'm going to rebase.

**Comment by clayne11** [comment]
Alright, it should be good to go now.

**Review by thasner — APPROVED** [review]

File: types/react-redux/index.d.ts [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
 type StatelessComponent<P> = React.StatelessComponent<P>;
-type Component<P> = ComponentClass<P> | StatelessComponent<P>;
+type Component<P> = React.ComponentType<P>;
 type ReactNode = React.ReactNode;
 type Store<S> = Redux.Store<S>;
 type Dispatch<S> = Redux.Dispatch<S>;
 type ActionCreator<A> = Redux.ActionCreator<A>;
 
+// Diff / Omit taken from https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-311923766
+type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
```
Comment (amcasey) [inline types/react-redux/index.d.ts]: These types seem rather complicated.  Would it make sense to add some sample uses to the test case below?

Comment (clayne11) [inline types/react-redux/index.d.ts]: I'm not opposed to doing that. It's definitely not trivial.

**Comment by mhegazy** [comment]
@clayne11 can you please resolve merge conflicts.

**Comment by coreh** [comment]
🎉  This is a significant improvement for the usability of the library.

One suggestion: Can we make the `dispatch` field optional in `DispatchProp`?

```
export interface DispatchProp<S> {
    dispatch?: Dispatch<S>;
}
```

That way I don't need to add a prop compatible with `dispatch: Dispatch<S>` to all my controlled components that don't specify mapDispatchToProps, (Sometimes there's really nothing to dispatch) but I can add it to the ones where I want to dispatch and it will be type checked.

**Comment by clayne11** [comment]
That makes sense to me. If you open a PR for that I'll approve it.

**Comment by 19majkel94** [comment]
@clayne11 
Looks like your types doesn't work with optional props, like the one with 3rd party HOC, e.g. redux-form:
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19989

**Pull request merged**